File: gunhoflash/clone-graph.py
class Node:
  def __init__(self, val = 0, neighbors = None):
    self.val = val
    self.neighbors = neighbors if neighbors is not None else []

class Solution:
  def cloneGraph(self, node: Node) -> Node:
    clones = [Node(i) for i in range(101)]
    isVisited = [False] * 101

    if node is None:
      return None

    visitingStack = [node]
    while len(visitingStack):
      now = visitingStack.pop()

      if isVisited[now.val]: continue
      isVisited[now.val] = True

      for neighbor in now.neighbors:
        if not isVisited[neighbor.val]:
          clones[now.val].neighbors.append(clones[neighbor.val])
          clones[neighbor.val].neighbors.append(clones[now.val])
          visitingStack.append(neighbor)

    return clones[node.val]

File: gunhoflash/clone-graph.ts
class GraphNode {
  val: number;
  neighbors: GraphNode[];

  constructor(val: number = 0, neighbors: GraphNode[] = []) {
    this.val = val;
    this.neighbors = neighbors;
  }
}

function cloneGraph(node: GraphNode | null): GraphNode | null {
  const clones = Array(101).fill(null).map((_, i) => new GraphNode(i))
  const isVisited = Array(101).fill(false);

  if (node === null) {
    return null;
  }

  const visitingStack = [node];
  while (visitingStack.length) {
    const now = visitingStack.pop()!;

    if (isVisited[now.val]) continue;
    isVisited[now.val] = true;

    now.neighbors
    .filter(neighbor => !isVisited[neighbor.val])
    .forEach(neighbor => {
      clones[now.val].neighbors.push(clones[neighbor.val]);
      clones[neighbor.val].neighbors.push(clones[now.val]);
      visitingStack.push(neighbor);
    });
  }

  return clones[node.val];
}
